src/EventStore.Core/Index/TableIndex.cs

```diff
@@ -647,28 +678,60 @@ public class TableIndex<TStreamId> : TableIndex, ITableIndex<TStreamId> {
 			if (endVersion < 0)
 				throw new ArgumentOutOfRangeException("endVersion");
 
-			var candidates = new List<IEnumerator<IndexEntry>>();
+			// 1. assemble results per table for memtables and ptables
+			// discard any results with 0 entries.
+			var resultsPerTable = new List<IList<IndexEntry>>(16);
 
 			var awaiting = _awaitingMemTables;
 			for (int index = 0; index < awaiting.Count; index++) {
-				var range = awaiting[index].Table.GetRange(hash, startVersion, endVersion, limit).GetEnumerator();
-				if (range.MoveNext())
-					candidates.Add(range);
+				var range = awaiting[index].Table.GetRange(hash, startVersion, endVersion, limit);
+				if (range.Count > 0)
+					resultsPerTable.Add(range);
 			}
 
 			var map = _indexMap;
 			foreach (var table in map.InOrder()) {
-				var range = table.GetRange(hash, startVersion, endVersion, limit).GetEnumerator();
-				if (range.MoveNext())
-					candidates.Add(range);
+				var range = table.GetRange(hash, startVersion, endVersion, limit);
+				if (range.Count > 0)
+					resultsPerTable.Add(range);
+			}
+
+			// 2. iterate through the per table results producing candidate enumerators
+			// if the results per table dont overlap then we can merge them much more quickly
+			// because they're already in order. this will be so in the common case of
+			// having no hash collisions.
+			var maybeInOrder = true;
+			var prevOldestVersion = long.MaxValue;
+			var totalEntryCount = 0;
+			var candidates = new List<IEnumerator<IndexEntry>>(resultsPerTable.Count);
+			foreach (var entries in resultsPerTable) {
+				var entriesEnumerator = entries.GetEnumerator();
+				entriesEnumerator.MoveNext();
+				candidates.Add(entriesEnumerator);
+
+				totalEntryCount += entries.Count;
+
+				if (maybeInOrder) {
+					var latestVersion = entriesEnumerator.Current.Version;
+
+					if (latestVersion >= prevOldestVersion) {
+						maybeInOrder = false;
+					}
+
+					var oldestVersion = entries[^1].Version;
+					prevOldestVersion = oldestVersion;
+				}
 			}
+			var areInOrder = maybeInOrder;
 
+			// 3. sort the entries by consuming the enumerators
+			//    removes duplicates
 			var last = new IndexEntry(0, 0, 0);
 			var first = true;
 
-			var sortedCandidates = new List<IndexEntry>();
+			var sortedCandidates = new List<IndexEntry>(totalEntryCount);
 			while (candidates.Count > 0) {
```

Comment:
We can additionally consider reversing the candidates list before entering the while loop to avoid O(n^2) operations when calling candidates.RemoveAt(0) here since it needs to shift the list every time: https://github.com/EventStore/EventStore/pull/3161/files#diff-1378774d17a4ef6f522262adffcb6e4fb2e29f275957abd040da1b53fda4b318R745

Reply:
👍 thats a neat solution, implemented